prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,152 @@ def check_for_location_conflicts(substations):
+    subcoord2subid = (
+        substations.round(rounding).groupby(["LONGITUDE", "LATITUDE"]).groups
+    )
+    print(
+        f"initial number of substations: {substations.shape[0]}\n"
+        f"substations with same location after rounding: "
+        f"{substations.shape[0] - len(subcoord2subid)}\n"
+    )
+
+    lines_coord_from = lines["COORDINATES"].map(
+        lambda x: (round(x[0][0], rounding), round(x[0][1], rounding))
+    )
+    lines_coord_to = lines["COORDINATES"].map(
+        lambda x: (round(x[-1][0], rounding), round(x[-1][1], rounding))
+    )
+    line2sub = lines_coord_from.to_frame(name="FROM").join(
+        lines_coord_to.to_frame(name="TO")
+    )
+
+    for endpoint in ["FROM", "TO"]:
+        line2sub[f"{endpoint}_SUB"] = line2sub[endpoint].apply(
+            lambda x: list(subcoord2subid[x]) if x in subcoord2subid else np.NaN
+        )
+
+    print(
+        f"initial number of lines: {lines.shape[0]}\n"
+        f"zero distance lines after rounding: "
+        f"{lines.shape[0] - line2sub['FROM'].compare(line2sub['TO']).shape[0]}\n"
+    )
+
+    from_sub_missing = line2sub["FROM_SUB"].isnull().values
+    to_sub_missing = line2sub["TO_SUB"].isnull().values
+
+    unique_endpoint_coord = set(
+        line2sub["FROM"].unique().tolist() + line2sub["TO"].unique().tolist()
+    )
+
+    n_two_sub_missing = sum(from_sub_missing & to_sub_missing)
+    n_one_sub_missing = sum(from_sub_missing ^ to_sub_missing)
+    n_unique_orphan_endpoint = len(unique_endpoint_coord - set(subcoord2subid))
+    n_unconnected_sub = len(set(subcoord2subid) - unique_endpoint_coord)
+
+    print(
+        f"there are {n_two_sub_missing} lines with two substations missing\n"
+        f"there are {n_one_sub_missing} lines with one substation missing\n"
+        f"there are {n_unique_orphan_endpoint} unique orphan endpoint\n"
+        f"there are {n_unconnected_sub} substations unconnected\n"
+    )
+
+    return line2sub
+
+
+def assign_substations_to_lines(substations, lines, line2sub):
+    """Locate closest substation of unconnected lines and update information
+    information of lines. Note that the ``line2sub`` data frame is modified in-place.
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param pandas.DataFrame line2sub: indices: table mapping lines to substations
+        as returned by :func:`map_lines_to_substations_using_coords`.
+    :return: (*pandas.DataFrame*) -- original lines data frame with updated
+        *SUB_1* and *SUB_2* columns, new *SUB_1_ID*, *SUB_2_ID* along with
+        *OTHER_SUB_1_ID* and *OTHER_SUB_2_ID* when multiple location are
+        associated to the same endpoint.
+    """
+    print("finding closest neighbor to unconnected lines")
+
+    # find closest neighbor of unconnected lines
+    subcoord2subid = substations.groupby(["LONGITUDE", "LATITUDE"]).groups
```

Comment:
Should we `round` here as well?

Reply:
I would say no. We can only gain at using the more precise located substations to find the closest neighbor.

Reply:
But this will give us a different group of substations which is not consistent with `map_lines_to_substations_using_coords` right? Would it be a problem that we use one grouping of substations when dealing with some lines and use a different grouping of substations when dealing with the rest of the lines?

Reply:
I don't think it matters. In `map_lines_to_substations_using_coords`, we identify the substations that have the exact same coordinates than the endpoints of the lines for a given rounding. In `assign_substations_to_lines` we assign substations to endpoints that have not been matched. If we don't round we will likely have a unique closest neighbor, if we round we will likely have multiple closest neighbor (with same location) and the algorithm will pick the first one it encounters in the list (see `powersimdata.utility.distance.find_closest_neighbor`). What do you think is better?

Reply:
I see your point. I was thinking the following situation:
Suppose we have raw sub list consists of: 
`coord1`, `sub_a`, "Substation"
`coord2`, `sub_b`, "Tap"
`coord3` `sub_c`,  "Deadend"
and line list consists of: 
`coord12`, `line_a`, 
`coord4`, `line_b`. 
After `groupby` with rounding, we have
`coord12`, [`sub_a`, `sub_b`], ["Substation", "Tap"]
`coord3`, `sub_c`, "Deadend"
Then we match `line_a` with `sub_a` via exact match of `coord12` and no match for `line_b`
Next we find nearest neighbor of `line_b` in the original sub list, and get `sub_b` with `coord2` (suppose dist(`coord4`, `coord2`) is the minimum). Then we will assign `sub_b` to `line_b` instead of `sub_a` in the other case. Not sure whether it matters and I don't have preferences between these two. Just bring this up for discussion.

Reply:
I guess I don't see the point of rounding and potentially having multiple closes neighbor if we don't have a selection criteria (voltage, number of lines connected to the substations, etc) in place. Otherwise, we take the closest one based on an surrealistic precision (what is implemented) or the one that appears first in the list as it is done in the closest neighbor algorithm (what you propose).

Reply:
My point in the above example is all matched lines will be assigned to `sub_a` if we follow consistent criteria in both cases (location matched directly + missing ones matched by nearest neighbor), however if we follow the current implementation, i.e. find nearest neighbor without rounding, we will have `sub_b` assigned to some of the lines. 

However, `sub_b` is unlikely to be an active candidate in the final substation list, given it is grouped together with `sub_a` which is a "SUBSTATION". Also I'm not sure at the end of day whether we will come up with universal substation IDs to replace the original ones, if so, are we going to assign different substation IDs to `OTHER_{endpoint}_SUB` as well?

Reply:
I understand your the difference between the 2 implementations. That said, there is no guarantee that if we round the coordinates leading to `sub_a = sub_b` that `sub_a` will be picked up. It will depend of their respective position in the substation data frame that is traversed row-wise to find the closest neighbor.

You are right that in the current implementation a line that is matched to a neighbor substation will have `OTHER_{endpoint}_SUB` set to `None`

Reply:
@danielolsen, what do you think? For lines that have no substations at their endpoint after `map_lines_to_substations_using_coords`, should we:
* Find closest neighbor in the original substation data frame (current implementation) or
* Find the closest location in the rounded substation data frame and pick one of the substation. `KDTree` returns a list of closest neighbors if they have same location.

Reply:
If we're saying that substations with coordinates that are close enough (i.e. the same after rounding) will be considered as one substation in the output, then I don't think it makes sense to differentiate between them when assigning lines to them. Whether we assign lines based on rounded coordinates (somehow picking the ID of the one substation that will remain), or assign lines based on the unrounded coordinates and then reassign any lines mapped to 'other' subs back to the main sub, the results should be equivalent, right?

Reply:
> assign lines based on the unrounded coordinates and then reassign any lines mapped to 'other' subs back to the main sub

@rouille In the current implementation, we don't reassign lines mapped to "other" subs back to the main sub on this track, do we?

Reply:
Not sure I understand. let me quickly summarize what the code is doing. We have two functions:
* `map_lines_to_substations_using_coords` that does the following
    * Round coordinates of substations and lines' endpoints
    * Map coordinates of lines' endpoints to coordinates of substations
    * Return a data frame where each line (index) can have at its endpoints a substation, a list of substation a Nan (no match):
```
>>> from prereise.gather.griddata.hifld.data_access.load import get_hifld_electric_power_transmission_lines, get_hifld_electric_substations
>>> from prereise.gather.griddata.hifld.data_process.transmission import map_lines_to_substations_using_coords
>>> sub = get_hifld_electric_substations("https://besciences.blob.core.windows.net/datasets/hifld/Electric_Substations_Jul2020.csv")
>>> line = get_hifld_electric_power_transmission_lines("https://besciences.blob.core.windows.net/datasets/hifld/Electric_Power_Transmission_Lines_Jul2020.geojson.zip")
>>> 
>>> line2sub = map_lines_to_substations_using_coords(sub, line, rounding = 3)
initial number of substations: 70857
substations with same location after rounding: 4464

initial number of lines: 71554
zero distance lines after rounding: 3314

there are 1172 lines with two substations missing
there are 2386 lines with one substation missing
there are 2584 unique orphan endpoint
there are 15946 substations unconnected
>>> line2sub.head(n=25)
                  FROM                  TO        FROM_SUB          TO_SUB
0   (48.188, -122.113)  (48.215, -122.028)  [55014]         [64253]       
1   (47.924, -122.104)  (47.925, -122.101)  [55018]         [56524]       
2   (47.838, -117.352)  (47.838, -117.41)   [55247]         [64264]       
3   (47.647, -121.907)  (47.648, -121.907)  [55295]         [64254]       
4   (46.873, -122.695)  (46.882, -122.7)    [55308]         [64255]       
5   (47.842, -122.3)    (47.842, -122.297)  [55359]         [64256]       
6   (47.799, -122.351)  (47.8, -122.351)    [55363]         [64257]       
7   (47.282, -119.555)  (47.282, -119.555)  [55376, 64258]  [55376, 64258]
8   (47.674, -122.158)  (47.675, -122.158)  [55436]         [64259, 64260]
9   (47.674, -122.158)  (47.675, -122.158)  [55436]         [64259, 64260]
10  (47.668, -122.196)  (47.668, -122.196)  [55437, 64261]  [55437, 64261]
11  (46.974, -122.827)  (46.976, -122.825)  [64262]         [55460]       
12  (46.977, -122.824)  (46.976, -122.825)  [64263]         [55460]       
13  (44.072, -122.967)  (44.071, -122.972)  [56648]         [56649]       
14  (45.628, -122.782)  (45.628, -122.786)  [56684]         [64265]       
15  (45.611, -122.773)  (45.614, -122.769)  [56685]         [61413, 61414]
16  (45.485, -122.749)  (45.486, -122.748)  [56717]         [60296]       
17  (44.292, -121.503)  (44.27, -121.493)   [56733]         [63254]       
18  (45.067, -117.864)  (45.05, -117.809)   [56736]         [64266]       
19  (45.904, -119.32)   (45.909, -119.323)  [64267]         [64314, 64315]
20  (45.904, -119.32)   (45.909, -119.323)  [64267]         [64314, 64315]
21  (45.917, -119.307)  (45.865, -119.3)    NaN             [64268]       
22  (45.699, -121.522)  (45.699, -121.522)  [56848, 60264]  [56848, 60264]
23  (45.699, -121.522)  (45.699, -121.521)  [56848, 60264]  [64269]       
24  (44.297, -120.886)  (44.298, -120.884)  [56944]         [64270]       
``` 
* `assign_substations_to_lines` that does:
    * Find closest substation for each line having one or two NaN value(s). CURRENTLY (this is what is discussed here) we look into the original substation data frame where each entry as a unique location. Only one substation is associated to the missing endpoint.
    * For line having multiple substation at their endpoint (e.g. line `7`), we keep the substation that has a `TYPE` in the following order of preference: SUBSTATION --> TAP --> RISER --> DEAD END. The other substations are saved and will are added  in the `OTHER_SUB_1_ID` and `OTHER_SUB_2_ID` of the returned lines data frame (see below)
    * A line data frame is returned with additional `SUB_1_ID`, `SUB_2_ID`, `OTHER_SUB_1_ID` and `OTHER_SUB_2_ID`. The `SUB_1` and `SUB_2` columns (name of substations) is overwritten with the substations' name that has been matched using the `SUB_1_ID`, `SUB_2_ID` columns. This data frame has the same number of rows than the original one.
    * We return a substation data frame with entries matching what can be found in `SUB_1_ID`, `SUB_2_ID`. Potentially, we could include substations in `OTHER_SUB_1_ID` and `OTHER_SUB_2_ID`.

Reply:
Thanks, this is clearer. There is a lot going on in the `assign_substations_to_lines` function and it was a little hard to keep straight in my head. If I understand properly, then for the returns of this function:
- within the lines data frame, each line maps to exactly one substation (either the one it was already mapped to, or a newly-mapped one), and if there were other substations that the line could have been mapped to, those are listed in the OTHER columns.
- the substation data frame only contains the substations that at least one line is mapped to.

Yeah?

Reply:
Correct.

Reply:
@rouille Very clear and it is consistent with my understanding. The line data frame is clean. Each line has two substations attached to both endpoints. I think the question is more for substation data frame, do we want any rows to be entries listed in `OTHER_SUB` columns in the line data frame, i.e. should there be lines connect to substations listed in `OTHER_SUB` columns?

Reply:
It would not hurt.

Reply:
Previously, I thought for the list of substations (`groupby` with rounding), we pick one according to the ranking as the unique substation entry and treat the rest ones in the list as functional affiliations. Hence, all the lines should have such unique entries as endpoints rather than affiliations. If that's not the case, I believe it's still okay. @danielolsen What do you think?

Reply:
I think I see what @BainanXia is getting at: if a line _already_ was mapped to a substation ID, and then that substation was 'collapsed' into a group of co-located substations, and the original substation was not chosen as the 'primary' substation, then I think we would want to re-map that endpoint from the 'secondary' substation to the 'primary' one. That way, every 'primary' substation ID maps to an entry in the substation data frame.

Reply:
I understand now. I will implement that

Reply:
This has been implemented in 71f13d9. This is what is happening now
* Call `map_lines_to_substations_using_coords`:
```
>>> from prereise.gather.griddata.hifld.data_access.load import get_hifld_electric_power_transmission_lines, get_hifld_electric_substations
>>> from prereise.gather.griddata.hifld.data_process.transmission import map_lines_to_substations_using_coords, build_transmission
>>> 
>>> sub = get_hifld_electric_substations("https://besciences.blob.core.windows.net/datasets/hifld/Electric_Substations_Jul2020.csv")
>>> line = get_hifld_electric_power_transmission_lines("https://besciences.blob.core.windows.net/datasets/hifld/Electric_Power_Transmission_Lines_Jul2020.geojson.zip")
>>> 
>>> line2sub = map_lines_to_substations_using_coords(sub, line, rounding = 3)
```
* Looking at line `87795` in data frame mapping lines to substations after rounding:
```
>>> line2sub.loc[87795]
FROM        (43.805, -95.175)
TO          (43.805, -95.175)
FROM_SUB       [42737, 52083]
TO_SUB                [42737]
Name: 87795, dtype: object
```
Since `FROM` = `TO`, then `FROM_SUB` = `TO_SUB` unless the location of substation `42737` matches exactly the location of the `TO` end of line `87795`.
* Looking at substation `42737`, we have:
```
>>> sub.loc[42737, "LATITUDE"]
43.805099
>>> sub.loc[42737, "LONGITUDE"]
-95.174541
```
* Looking at the coordinates of line `87795`:
```
>>> line.loc[87795, "COORDINATES"]
[[43.804914316000065, -95.17504414599995], [43.80509324500008, -95.17512502199997], [43.805211787000076, -95.17460879499998], [43.805099, -95.174541]]
```
The `FROM` coordinates (first list) is slightly different than the `TO` coordinates (last list) and the latter matches exactly the coordinates of substation `42737`

I think this is what we want.

Reply:
Playing around with this:
```python
>>> from prereise.gather.griddata.hifld.data_process.transmission import build_transmission
>>> lines, substations  = build_transmission(method="line2sub", kwargs={"rounding": 2})
[output truncated for brevity]
>>> len(lines)
71554
>>> len(substations)
54487
>>> lines["SUB_1_ID"].isin(substations.index).all()
True
>>> lines["SUB_2_ID"].isin(substations.index).all()
True
>>> primary_substations = set(lines["SUB_1_ID"].unique()) | set(lines["SUB_2_ID"].unique())
>>> len(primary_substations)
43285
>>> len(substations.loc[primary_substations].groupby(["LATITUDE", "LONGITUDE"]))
43285
>>> len(substations.loc[primary_substations, ["LATITUDE", "LONGITUDE"]].round(2).groupby(["LATITUDE", "LONGITUDE"]))
42671
```
It seems that there are no 'primary' substations with the same lat/lon, but there about 600 that overlap within the rounding precision at rounding=2, or about 180 when we round to 3 decimal places (50708 total, 50529 after grouping).

Reply:
If I'm understanding correctly, here is what's going on:

1. Map lines to substations after rounding, some endpoints of lines are assigned a list of substations since they are grouped due to rounding.
2. For those endpoints above, look at the coordinates before rounding and check if there is an exact match in the list, if so, drop the rest elements in the list, if not, pick the primary substation according to the ranking then drop the rest elements in the list (later).
3. For the rest endpoints with no match in 1, assign the closest neighbor to it.
4. Recreate the list of substations based on the resulting line data frame, any substations matched to a line (with rounding) are kept and the rest ones are dropped.

Then I'm wondering whether we could simplify the process as following:

1. Map endpoints of lines to substations based on exact match of coordinates .
2. Map no-exact-match endpoints of lines to substations with rounding.
3. Map remaining endpoints of lines to nearest substations.

Further thoughts regarding new substation data frame:

1. Shall we keep row entries to be primary substations, i.e. substations with lines directly connecting to them. In this way` len(primary_substations)` == `len(substations)`
2. Move `OTHER_SUB` columns from line data frame to substation data frame, since this might be cleaner without duplicated entries (currently, if multiple endpoints are mapped to a same group of substations with rounding, they will have duplicated value in the `OTHER_SUB` columns.

All draft thoughts. Open to discussions.

Reply:
@BainanXia, I believe that the first implementation and the second will lead to the same outcome. 

Regarding the difference in length between the output `substations` and @danielolsen's `primary_substations`, it comes from the fact that we include substations listed in `OTHER_FROM_SUB` and `OTHER_TO_SUB` into the output substations. Would you prefer to get to not include them?

@BainanXia, I don't understand your second point. How will we connect the line to substations listed in the `OTHER_FROM_SUB` and `OTHER_TO_SUB` columns if these are moved into the substations data frame?

Reply:
@rouille 
- Agree. The two implementations should lead to the same outcome. I bring it up just for simpler logic flow.
- Yes. Following our current setup in the `usa_tamu` model, all substations (row entries) are primary substations. I lean to not including them as separate row entries in the data frame.
- I was thinking instead of `OTHER_FROM_SUB` and `OTHER_TO_SUB` in line data frame, we just maintain an `OTHER_SUB` column in substation data frame.

Reply:
@BainanXia I think there is still value in reporting the 'secondary' substation rows. E.g. if we want to augment our network with underground lines that we don't think are present in the HIFLD data set, it is helpful to know which 'primary' substations have a secondary `RISER` substation.

Reply:
@danielolsen Agree. What I'm proposing is to keep them in the substation data frame instead of line data frame. What do you think?

Reply:
Ah, right, I had meant to reply to that point as well and agree with it. Secondary substation IDs seem more related to substations themselves than lines which connect to them.